src/Campaigns/Migrations/MigrateFormsToCampaignForms.php

```diff
@@ -29,66 +26,153 @@ public static function id(): string
+            }
+        });
+    }
+
+    /**
+     * @unreleased
+     */
+    protected function getFormData(): array
+    {
+        $query = DB::table('posts', 'forms')
+            ->select(
+                ['forms.ID', 'id'],
+                ['forms.post_title', 'title'],
+                ['forms.post_status', 'status'],
+                ['forms.post_date', 'createdAt']
+            )
+            ->where('forms.post_type', 'give_forms');
+
+        $query->select(['formmeta.meta_value', 'settings'])
+            ->join(function (JoinQueryBuilder $builder) {
+                $builder
+                    ->leftJoin('give_formmeta', 'formmeta')
+                    ->on('formmeta.form_id', 'forms.ID');
+            })
+            ->where('formmeta.meta_key', 'formBuilderSettings');
+
+        // Exclude forms already associated with a campaign (ie Peer-to-peer).
+        $query->join(function (JoinQueryBuilder $builder) {
+                $builder
+                    ->leftJoin('give_campaigns', 'campaigns')
+                    ->on('campaigns.form_id', 'forms.ID');
+            })
+            ->whereIsNull('campaigns.id');
+
+        // Exclude forms with an `upgraded` status, which are archived.
+        $query->where('forms.post_status', 'upgraded', '!=');
+
+        return $query->getAll();
     }
 
     /**
      * @unreleased
+     * @return array [{formId, campaignId, migratedFormId}]
      */
-    public function createParentCampaignForDonationForm(DonationForm $form)
+    protected function getUpgradedFormData(): array
     {
-        $campaign = Campaign::create([
-            'type' => CampaignType::CORE(),
-            'title' => $form->title,
-            'shortDescription' => $form->settings->formExcerpt,
-            'longDescription' => $form->settings->description,
-            'logo' => $form->settings->designSettingsLogoUrl,
-            'image' => $form->settings->designSettingsImageUrl,
-            'primaryColor' => $form->settings->primaryColor,
-            'secondaryColor' => $form->settings->secondaryColor,
-            'goal' => (int) $form->settings->goalAmount,
-            'status' => $this->mapFormStatusToCampaignStatus($form->status),
-            'startDate' => new DateTime($form->settings->goalStartDate),
-            'endDate' => new DateTime($form->settings->goalEndDate),
-        ]);
+        return DB::table('posts', 'forms')
+            ->select(['forms.ID', 'formId'], ['campaign_forms.campaign_id', 'campaignId'])
+            ->attachMeta('give_formmeta', 'ID', 'form_id', 'migratedFormId')
+            ->join(function (JoinQueryBuilder $builder) {
+                $builder
+                    ->rightJoin('give_campaign_forms', 'campaign_forms')
+                    ->on('campaign_forms.form_id', 'forms.ID');
+            })
+            ->where('forms.post_type', 'give_forms')
+            ->where('forms.post_status', 'trash', '!=')
```

Comment:
Not sure if we should exclude forms with the trash status. I know that our list tables don't provide a way to restore items with this status, but in the legacy list view, this is possible. So, thinking about the feature flag approach that we discussed previously, maybe some customers still have access to that legacy list view table which will allow them to restore these forms that will not be associated with any campaign in that scenario.

![image](https://github.com/user-attachments/assets/b1b13fa1-e883-48c8-aec2-0970234cbc50)

Reply:
Good point, @glaubersilva. This also brings up the direction that "every form is part of a campaign", even a deleted for would need a campaign.

Reply:
Resolved, see [6146be5](https://github.com/impress-org/givewp/pull/7532/commits/6146be5b853c6048987b4147e1e690e7932ff502).